Refine single-column Table to act as a Column

radeusgd opened the pull request:
### Pull Request Description

- Closes #12137

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

radeusgd committed:
use `from_java_table` instead of `Table.Value` to get a single 'entry point'

radeusgd committed:
rename

radeusgd committed:
add single column refinement

radeusgd committed:
workaround for IR cache bug

radeusgd committed:
add tests for in-memory

radeusgd committed:
one more test case

radeusgd committed:
add analogous DB_Table tests

radeusgd committed:
fixes

radeusgd committed:
implement for DB_Table

radeusgd committed:
some alignment/fixes to tests

radeusgd committed:
avoid warning in test

radeusgd committed:
add tests for single column + single value

radeusgd committed:
copy paste to get single-column single-value table working

radeusgd committed:
given we aren't re-using column logic, we should test all cases to ensure the copy-pasted code is not buggy...

radeusgd committed:
remove Text.from Table conversion and replace with explicit `to_delimited` method

radeusgd committed:
test

radeusgd committed:
workaround for https://github.com/enso-org/enso/issues/12180

radeusgd committed:
workaround for https://github.com/enso-org/enso/issues/12185 - stop using `is_nothing`

radeusgd committed:
another workaround for https://github.com/enso-org/enso/issues/12180

radeusgd committed:
fix precedence in test

radeusgd committed:
add test

radeusgd committed:
workaround for `.catch` bug, but can stay

radeusgd committed:
replace usages of removed `Text.from Table` conversion with `to_delimited`

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/12137-table-as-single-column

radeusgd commented on distribution/lib/Standard/Database/0.0.0-dev/src/Connection/Connection.enso:
```diff
@@ -547,8 +547,9 @@ private _make_table_for_name connection name alias internal_temporary_keep_alive
            access any table already present in the database.
         DB_Table_Module.make_table connection name columns ctx on_problems=Problem_Behavior.Report_Warning
-    result.catch SQL_Error _->
+    if result.is_error then
         Error.throw (Table_Not_Found.Error name)
+    result
```
Since #11777 we can simplify the error handling code, let's use it :)

radeusgd commented on distribution/lib/Standard/Table/0.0.0-dev/src/Table.enso:
```diff
@@ -3923,11 +3929,19 @@ type Table
         data = self.columns.map c->("[" + c.name.pretty + ", " + c.to_vector.pretty + "]") . join ", "
         "Table.new [" + data + "]"
 
+    ## Converts all table contents to a text value in delimited format.
+
+       Arguments:
+       - format: Allows to customize the delimiter and other settings of the
+         format. Defaults to tab-separated values.
+    to_delimited self (format:Delimited_Format = ..Delimited '\t') -> Text =
+        Delimited_Writer.write_text self format
```
We have agreed that `Text.from Table` conversion wasn't that useful and so we are turning it into a method on `Table` instead.

Does that name seem good or any suggestions for a better one?

jdunkerley replied:
Matches with `to_xml`

jdunkerley commented on distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Type_Refinements/Table_Refinements.enso:
```diff
@@ -0,0 +1,38 @@
+refine_table (table : Table) =
+    if is_single_column table . not then table else
+        column = table.at 0
+        # This should be consistent with `Column_Refinements.refine_column` - the code needs to be copied because we need to spell out all the types.
```
Probably one we need to talk over with Jaroslav to see if we can do anything better longer term.

GregoryTravis commented on distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Type_Refinements/Table_Refinements.enso:
```diff
@@ -0,0 +1,38 @@
+            Value_Type.Time -> (table : Table & Column & Time_Of_Day)
+            Value_Type.Date_Time True -> (table : Table & Column & Date_Time)
+            Value_Type.Decimal _ scale ->
+                is_integer = scale == 0
```
What about `scale < 0`? It can still be an integer, so would `Table & Column & Integer` be the right type in that case?

radeusgd replied:
Good point! I forgot it can be negative.

radeusgd replied:
Updated in bc718a20

GregoryTravis commented on distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Read_Many_Helpers.enso:
```diff
@@ -77,26 +77,26 @@ _merge_input_and_tables (input_table : Table) (tables_for_rows : Vector Read_Man
         multiplicated_inputs = duplicate_rows input_table counts
         Runtime.assert (unified_data.row_count == multiplicated_inputs.row_count)
-        Runtime.assert (unified_metadata.is_nothing || (unified_metadata.row_count == unified_data.row_count))
+        Runtime.assert ((Nothing == unified_metadata) || (unified_metadata.row_count == unified_data.row_count))
```
Why is this change needed?

radeusgd replied:
Due to #12185.

Essentially - `unified_metadata` is a `Table`. But if it contains 1 column, it also now is a `& Column`. And due to #12185 bug, when calling `unified_metadata.is_nothing`, instead of dispatching to the `Any.is_nothing` implementation, it dispatches to `Column.is_nothing` and returns `Column` instead of `Boolean` - the `assert` then crashes as it expects a boolean.

`Nothing == x` goes around this issue and checks the **is this a Nothing value** property correctly.

GregoryTravis commented on test/Table_Tests/src/Database/Common/Single_Column_Table_Spec.enso:
```diff
@@ -0,0 +1,48 @@
+            t1 = setup.table_builder [["A", [23]]]
+            t1.should_be_a DB_Table
+            (t1:DB_Column).name.should_equal "A"
+            Test.expect_panic Type_Error (t1:Integer).to_vector
```
I am confused by this test. `to_vector` is not a method of `Integer`, so I would expect a method-not-found error here.

radeusgd replied:
Right, it is a typo, thanks for noticing!

radeusgd replied:
The test was still succeeding because it `t1:Integer` is (expectedly) failing with `Type_Error`. So it never reached the `to_vector` method call... But it shouldn't have been there, I removed it.

AdRiley commented on distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso:
```diff
@@ -90,6 +89,14 @@ type DB_Table
+    ## The internal constructor used to construct a DB_Table instance.
+
+       It can perform some additional operations, like refining the type,
+       so it should always be preferred over calling `DB_Table.Value` directly.
```
How would I know this without reading this comment? Guess we need some sort of convention for module private constructors similar to what we have for module priavte methods

radeusgd replied:
Good point, I will add a comment in `Value` constructor that it should not be used directly.

radeusgd replied:
We can also consider changing the name of the constructor from `Value` to something like `Internal_Value` or `Internals` - to make it more apparent that this is some internal thing that should only be used in limited places. What do you think?

AdRiley reviewed (approved):
Seeing it in use has led me to a new question. (And I think this is the question that has been in my head all week trying to get out. :))

If the types are hidden then how are they any use? And I think the answer we have been saying is: well they can be cast to the hidden types. And the GUI will cast them to the hidden types. But how will the GUI know it can cast them to the hidden types if they are hidden?

radeusgd commented:
> But how will the GUI know it can cast them to the hidden types if they are hidden?

@AdRiley The idea is that the GUI should be able to 'extract' even the 'hidden' part of the type.

As far as I understand it was supposed to be handled by PR #11583 (and perhaps some followup changes). I think the idea is that the `types` sent to the IDE include not only the 'visible' but also the 'hidden' part. Now - I think the message currently does not distinguish which part is which - it just sends _all_ the types a value has. Perhaps we will need to expand the protocol to include such a distinction (e.g. if we only wanted to insert the casts for hidden types, as including it for visible ones may be redundant).

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/12137-table-as-single-column

# Conflicts:
#	test/Base_Tests/src/Semantic/Multi_Value_As_Type_Refinement_Spec.enso

radeusgd committed:
CR: single helper for is decimal integer

radeusgd committed:
CR: comments about not using internal constructors

radeusgd committed:
CR: fix typo in test

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/12137-table-as-single-column

radeusgd committed:
name matches as per new conventions

radeusgd committed:
more: name matches as per new conventions

radeusgd committed:
reorder: a Table may also be a Column, so we need to check Table first because original type should be preferred

Note: a Column may be _converted_ to Table but is not a Column&Table, and conversions are not checked in case-of, so the other direction is not a problem
